### PR TITLE
Fix for CodeGenerator.gerenate() and FeatureExtractor._add_noops() if we remove labels with low support

### PR DESCRIPTION
**Pull request by zurk**
Second attempt to fix missing `vnode.y`

Issue description: https://github.com/src-d/style-analyzer/pull/408#issue-237166170
and related issue: https://github.com/src-d/style-analyzer/issues/406

### CODE REVIEW

**Comment by zurk** [comment]
`update_mappings` is called from two places, so it is better to keep it.

**Comment by m09** [comment]
sorry I should have been clearer: I think we can centralize what we do in update_mappings in _remove_labels_with_low_support (and rename it _extract_labels).

**Comment by zurk** [comment]
let's name it _extract_labels_mappings to be more precise?

**Comment by m09** [comment]
the mappings are implementation details, the goal is to extract labels. But up to you it's really minor.

**Comment by zurk** [comment]
just `_extract_labels` confuses me and I expect some kind of output with labels from a call like `_extract_labels(vnodes)`.

**Comment by m09** [comment]
sure but I guess there is the same problem with _extract_labels_mappings. So maybe _compute_labels ? Anyway, any name will do really :)

File: lookout/style/format/feature_extractor.py [hunk 1]
```diff
@@ -261,14 +261,18 @@ def extract_features(self, files: Iterable[File], lines: List[List[int]]=None) \
...
             self._fill_vnode_parents(file_parents, file_vnodes, uast, vnode_parents)
 
-        self._log.debug("Parsed %d vnodes", sum(len(vn) for vn, _, _ in parsed_files))
+        vnodes_parsed_number = sum(len(vn) for vn, _, _ in parsed_files)
+        self._log.debug("Parsed %d vnodes", vnodes_parsed_number)
         # filter composite labels by support
-        if index_labels and self.cutoff_label_support > 0:
-            self._remove_labels_with_low_support([vn for vns, _, _ in parsed_files for vn in vns])
+        if index_labels:
+            self._compute_labels_mappings([vn for vns, _, _ in parsed_files for vn in vns])
```
Comment (m09) [inline lookout/style/format/feature_extractor.py]: we can go back to iterator :)

File: lookout/style/format/feature_extractor.py [hunk 2]
```diff
@@ -728,21 +717,30 @@ def _parse_file(self, contents: str, root: bblfsh.Node, path: str) -> \
...
             if vnode.y is not None:
-                label_support[vnode.y] += 1
-        unsupported = {k for k, v in label_support.items() if v < self.cutoff_label_support}
-        for vnode in vnodes:
-            if vnode.y in unsupported:
-                vnode.y = None
-        self._log.debug("Removed %d/%d labels by support %d", len(unsupported), len(label_support),
+                support[vnode.y] += 1
+
+        for class_seq in sorted(support, key=lambda l: -support[l]):
```
Comment (m09) [inline lookout/style/format/feature_extractor.py]: O(n log n) instead of O(n) and less readable

**Comment by zurk** [comment]
Are you fine if I ask Vadim to review and merge it?

**Comment by m09** [comment]
sure

File: lookout/style/format/feature_extractor.py (hunk 2 above)
Comment (m09) [reply lookout/style/format/feature_extractor.py]: if it was not an optimization but a feature it's fine though

Comment (zurk) [reply lookout/style/format/feature_extractor.py]: It is small array ~100 of items, so does not really affect performance.
I just think it is a good idea to have this order: from the most frequent label to the least frequent.

Comment (m09) [reply lookout/style/format/feature_extractor.py]: yup I agree

File: lookout/style/format/feature_extractor.py [hunk 3]
```diff
@@ -728,21 +717,30 @@ def _parse_file(self, contents: str, root: bblfsh.Node, path: str) -> \
...
```
Comment (vmarkovtsev) [inline lookout/style/format/feature_extractor.py]:
```python
self.labels_to_class_sequences = [key for key, val in support.items() if val >= self.cutoff_label_support]
```

Comment (m09) [reply lookout/style/format/feature_extractor.py]: @vmarkovtsev: as discussed above sorting the labels by support makes sense to identify quickly if the label at hand is common or not

File: lookout/style/format/feature_extractor.py (hunk 2 above)
Comment (m09) [reply lookout/style/format/feature_extractor.py]: The for loop does nothing

**Pull request merged**